docs: add modifier documentation (#182)
Quantinuum/guppy-docs: sphinx/language_guide/modifiers.md

---
file_format: mystnb
kernelspec:
  name: python3
---

> # Modifiers

Reviewer: This page title will appear in the sidebar of the language guide. I think it will not be obvious to the reader what  "Modifiers" are. Perhaps its worth using one of the following.

* `# Modifiers - Control and Dagger`
* `# Controlling and Daggering quantum operations`

> Modifiers transform a block of quantum operations. They make it possible to express controlled and inverse operations without defining a separate function for each variant.

Reviewer: Another point is that we should only have one title level heading. Headings with a single `#` are rendered in the sidebar like this

<img width="226" height="371" alt="Image" src="https://github.com/user-attachments/assets/bc5f75ee-fb43-433e-aa53-404b7ae85526" />

This is not what we want as control, dagger, function flags are all subsections of the "Modifiers" page

Reviewer: Headings should also always be consecutive

> Modifiers transform a block of quantum operations. They make it possible to express controlled and inverse operations without defining a separate function for each variant.

Reviewer 2: I would prefer more here - I think it's important to highlight that modifiers are capable of changing the typoe signature (e. g U -> ctrl -U). What is the aim of modifiers? Are they there to improve composability? Are they there to defer to the compiler instead of user defns?

## Syntax

Use modifiers in a `with` statement. Multiple modifiers may be combined or nested.

```{code-cell} ipython3
from guppylang import guppy
from guppylang.std.builtins import control, dagger
from guppylang.std.quantum import h, qubit

@guppy

> def controlled_inverse(c: qubit, q: qubit) -> None:

Reviewer: I think the Hadamard gate is not a good example for the inverse as it is in fact equal to its own inverse $H = H^\dagger$. Maybe try the $S$ gate instead?

>     with control(c), dagger:

Reviewer 2: before you do a compound/nested example, please just do the simplest. Definitely a case where playpond integration would be rad because I'd like to see some output changing with and without e.g. daggering an operation

        h(q)

> controlled_inverse.check()

Reviewer 2: Follow up sentence after this snippet please before moving on to stuff that doesn't work 

'Here we take the S gate, and modify it with control and dagger. This gate has a known form of control and daggering, and the operations are compatible. Therefore, when we check the program (compile) the code passes. When applied this functions acts as a C-S^\dag  gate'

```

> The body has access to variables from its enclosing scope, but it cannot take ownership of them. For instance, the following programs is rejected:

Reviewer 2: The errors here are great but I'd prefer more explicit naming here than 'body' -- there is the original operator and the modified version. The modifier cannot take q from the scope of the original function

Reviewer 2: (and therefore it cannot discard or measure it?)

Reviewer 2: (Also why? This is so types/resources are clean?)

```{code-cell} ipython3
---
tags: [raises-exception]
---
from guppylang.std.builtins import owned
from guppylang.std.quantum import discard

@guppy.declare(daggerable=True)
def consume(q: qubit @ owned) -> None: ...

@guppy
def cannot_take_ownership(q: qubit) -> None:
    with dagger:
        consume(q)

cannot_take_ownership.check()
```

Moreover, assignments in a modifier block are local to that block, including assignments that reuse an outer name. In the following examples, `denominator` is not available outside the `with dagger:` block as well as `outer_var`, which is assigned outside the block but it is reassigned inside the block.

```{code-cell} ipython3
---
tags: [raises-exception]
---
from guppylang.std.quantum import angle, rx

@guppy
def local_assignment(q: qubit) -> None:
    outer_var = 1
    with dagger:
        denominator = 4
        outer_var = 2
        rx(q, angle(outer_var / denominator))
    # `denominator` is not available here.
    rx(q, angle(1 / denominator))

local_assignment.check()

> ```

Reviewer 2: prefer not to have dense snippets like this with no explanation between and with the same named functions (just append? e.g. local_assignment_fail and local_assignment_fail_2)

```{code-cell} ipython3
---
tags: [raises-exception]
---
@guppy
def local_assignment(q: qubit) -> None:
    outer_var = 1
    with dagger:
        denominator = 4
        outer_var = 2
        rx(q, angle(outer_var / denominator))
    # `outer_var` is not available here either
    rx(q, angle(outer_var / 4))

local_assignment.check()
```

# Control

`with control(c):` applies the operations in its body only when the control qubit `c` is in $\ket{1}$. With controls $c_1, \ldots, c_n$, it applies the operation when all controls are in $\ket{1}$:

$$
\operatorname{C}^n(U)\ket{c_1\ldots c_n}\ket{\psi} =

> \begin{cases}

Reviewer: I think we should use a zero indexed definition of $C^n(U)$ to be consistent with our convention for Guppy arrays and Python in general

$$
|c_0 \ldots c_{n-1} \rangle U|\psi\rangle
$$

\ket{c_1\ldots c_n}U\ket{\psi} & \text{if } c_1 = \cdots = c_n = 1, \\
\ket{c_1\ldots c_n}\ket{\psi} & \text{otherwise.}
\end{cases}
$$

You can pass individual qubits or an array of qubits. Array elements may also be controls.

```{code-cell} ipython3
from guppylang import array
from guppylang.std.quantum import x
from guppylang.std.builtins import nat

@guppy
def c2x(controls0: qubit, controls1: qubit, target: qubit) -> None:
    with control(controls0, controls1):
        x(target)

@guppy
def cnx[n: nat](controls: array[qubit, n], target: qubit) -> None:
    with control(controls):
        x(target)

@guppy
def main(c0: qubit, c1: qubit, t: qubit) -> None:
    
    c2x(c0, c1, t)

    controls = array([c0, c1])
    t2 = qubit()
    cnx(controls, t2)

cnx.check()
```

> ## Control flow

Reviewer 2: Annoying, but probably best to label this as 'Classical control' every time because of the competing namespace of control in this document

> When a controlled block contains control flow, the control is pushed to every quantum operation produced by the branch or loop. Evaluating the classical condition and loop bounds is not controlled. For instance, the following two programs are equivalent:

Reviewer 2: the guppy logic here is so clean and demonstrative you could maybe reverse this section - 'These are equivalent.. this is because classical conditioning and loops are not part of the with control modification'

Author: I see, but I'm not sure how to organize the session 🤔

```{code-cell} ipython3
from guppylang.std.quantum import angle, rx

@guppy
def control_assignment(c: qubit, q: qubit) -> None:
    with control(c):
        theta = angle(1 / 4)
        rx(q, theta)

@guppy
def pushed_control_assignment(c: qubit, q: qubit) -> None:
    theta = angle(1 / 4)
    with control(c):
        rx(q, theta)

control_assignment.check()
pushed_control_assignment.check()
```

## Forbidden operations

Control blocks allows classical operations since they can be evaluated without affecting the quantum state. However, they cannot allocate, measure, reset, or discard qubits, since these operations have quantum effects, but they are not controllable.

```{code-cell} ipython3
---
tags: [raises-exception]
---
@guppy
def allocation_in_control(c: qubit) -> None:
    with control(c):
        q = qubit()

allocation_in_control.check()
```

# Dagger

`with dagger:` applies the inverse of its body. If the body performs $U_1$ followed by $U_2$, the dagger block performs $U_2^\dagger$ followed by $U_1^\dagger$.

```{code-cell} ipython3
from guppylang.std.quantum import s

@guppy

> def sdg(q: qubit) -> None:

Reviewer: Maybe you could apply more than one gate inside this function to illustrate the point you made about $U_1$ and $U_2$ above. 

Perhaps $S$ followed by $X$? It would also make more sense as we already have a native `sdg` gate in `guppylang.std.quantum`.

    with dagger:
        s(q)

sdg.check()
```

## Classical assignments in dagger blocks

Similar to the control modifier, dagger reverses only the quantum computation. Classical assignments keep their source order, while the quantum operations are inverted and reversed.

```{code-cell} ipython3
@guppy
def invert_two_gates(q: qubit) -> None:
    with dagger:
        h(q)
        s(q)

invert_two_gates.check()
```

The resulting quantum operations are equivalent to:

```qasm

> sdag q;

Reviewer: Is `sdag` valid qasm? I thought it was `sdg`? Worth checking

        if flag:
            h(q)

branch_in_dagger.check()
```

```{code-cell} ipython3
---
tags: [raises-exception]
---
@guppy
def loop_in_dagger(q: qubit) -> None:
    with dagger:
        for _ in range(2):
            h(q)

loop_in_dagger.check()
```

```{code-cell} ipython3
---
tags: [raises-exception]
---
from guppylang.std.builtins import output

@guppy
def output_in_dagger(q: qubit) -> None:
    with dagger:
        output("value", True)

output_in_dagger.check()
```
## Combining dagger and controls

Modifiers compose. This example is a doubly controlled inverse of `h` followed by `s`:

```{code-cell} ipython3
@guppy
def doubly_controlled_inverse(c0: qubit, c1: qubit, q: qubit) -> None:
    with control(c0):
        with dagger:
            with control(c1):
                h(q)
                s(q)

doubly_controlled_inverse.check()
```

Resolving one modifier at a time gives this sequence:

```text
Source:              h q;                    s q;
Push control c1:     ctrl @ h c1, q;         ctrl @ s c1, q;
Resolve dagger:      ctrl @ sdag c1, q;      ctrl @ h c1, q;
Push control c0:     ctrl(2) @ sdag c0, c1, q;
                     ctrl(2) @ h c0, c1, q;
```

> ## Conjugation box

Reviewer: I would avoid any "Box" terminology as "Box" is very much a pytket feature rather than a Guppy feature. I'd go with 

1. Just "Conjugation"
2. "Conjugation pattern"
3. "Conjugation subroutine"

Many unitaries have a compute--action--uncompute form:

$$
U = V A V^\dagger.
$$

The first part computes a basis change, `A` performs the action, and the final part uncomputes the basis change. This is a conjugation box.

> For a controlled conjugation box, only the central action needs to be controlled:

Reviewer: ```suggestion
For a controlled conjugation, only the central action $A$ needs to be controlled:
```

---
file_format: mystnb
kernelspec:
  name: python3
---

# Controlling and Daggering quantum operations

Modifiers transform a block of quantum operations.
They automatically produce controlled and inverse versions of quantum operations. They apply to a single gate, a block containing many gates, or a function, so you can write an operation once and reuse its controlled or daggered form. Guppy generates the transformed operations, freeing you from defining and maintaining each variant by hand.

The modifier changes the underlying gates. For example, controlling a block adds the control to every gate it produces; effectively, a function `f(q)` becomes a controlled operation `ctrl-f(c, q)` with an additional control-qubit input. A dagger block reverses the gate order and replaces each gate with its inverse.

## Syntax

> Use modifiers in a `with` statement.

Reviewer: ```suggestion
We can make use of modifiers inside a context manager using the `with` keyword.
```

> Start with the simplest form: a single modifier on a single operation.

Reviewer: ```suggestion
Let’s start with a simple example of a controlled single-qubit operation. We will realize this with the `control` modifier. 
```

```{code-cell} ipython3
from guppylang import guppy
from guppylang.std.builtins import control, output
from guppylang.std.quantum import qubit, x, h, measure

@guppy

> def controlled_x() -> None:

Reviewer: I wouldn't use CX here. In general people should be using modifiers for gates which do not exist in `std.quantum`

    c = qubit()
    q = qubit()
    h(q)
    with control(c):
        x(q)

>     bit_c =measure(c)

Reviewer: Nit:

```suggestion
    bit_c = measure(c)
```

    bit_q = measure(q)
    output("result", array(bit_c.read(), bit_q.read()))

controlled_x.emulator(n_qubits=2).with_shots(100).run().collated_counts()
```
Here we can observe that the `x` operation is applied only when the control qubit `c` is in the $\ket{1}$ state.

```{code-cell} ipython3
from guppylang.std.builtins import output
from guppylang.std.quantum import angle, h, measure, qubit, rx

@guppy
def rotate_then_dagger() -> None:
    q = qubit()
    a = angle(1 / 4)
    rx(q, a)
    with dagger:
        rx(q, a)
    bit = measure(q)
    output("result", bit.read())
    

rotate_then_dagger.emulator(n_qubits=2).with_shots(100).run().collated_counts()
```
Since we are applying a rotation followed by its inverse, the qubit is always measured in the $\ket{0}$ state.

Multiple modifiers may also be combined or nested.

```{code-cell} ipython3
from guppylang.std.quantum import s, qubit

@guppy
def controlled_inverse(c: qubit, q: qubit) -> None:
    with control(c), dagger:
        s(q)

controlled_inverse.check()
```

> Here we take the S gate and modify it with control and dagger. This gate has a known form of control and daggering, and the operations are compatible. Therefore, when we check the program (compile), the code passes. When applied, this function acts as a C-S^\dag gate

Reviewer: ```suggestion
Here we take the $S$ gate and modify it with control and dagger. This gate has a known form of control and daggering, and the operations are compatible. Therefore, we can successfully type check the program. When applied, this function acts as a $CS^\dagger$ gate
```

Reviewer: I am not sure what this sentance means

"This gate has a known form of control and daggering, and the operations are compatible"

Is this just saying that the gate is unitary and therefore we can add a control to it?

Control blocks allow classical operations since they can be evaluated without affecting the quantum state. However, they cannot allocate, measure, reset, or discard qubits, since these operations have quantum effects, but they are not controllable.

```{code-cell} ipython3
---
tags: [raises-exception]
---
@guppy
def allocation_in_control(c: qubit) -> None:
    with control(c):
        q = qubit()

allocation_in_control.check()
```

## Dagger

`with dagger:` applies the inverse of its body. If the body performs $U_1$ followed by $U_2$, the dagger block performs $U_2^\dagger$ followed by $U_1^\dagger$.

```{code-cell} ipython3
from guppylang.std.quantum import s

@guppy
def sx_dg(q: qubit) -> None:
    with dagger:
        s(q)
        x(q)

sx_dg.check()
```

### Classical assignments in dagger blocks

Similar to the control modifier, dagger reverses only the quantum computation. Classical assignments keep their source order, while the quantum operations are inverted and reversed.

```{code-cell} ipython3
@guppy
def invert_two_gates(q: qubit) -> None:
    with dagger:
        h(q)
        s(q)

invert_two_gates.check()
```

The resulting quantum operations are equivalent to:

> ```qasm

Reviewer: Is the qasm annotation needed here?

sdg q;
h q;
```

Classical operations remain in order even though the quantum operations are reversed:

```{code-cell} ipython3
from guppylang.std.quantum import rz

@guppy

> def invert_rotations(q: qubit) -> None:

Reviewer: This example snippet seems to not execute without error on C.I.

<img width="726" height="554" alt="Image" src="https://github.com/user-attachments/assets/ebddcb83-731c-4f22-a3fd-70aae4316e2f" />

Author: After the cange in rc2 to limit the presence of classical computation in dagger I think this example should be removed. I wuold reintroduce this when we are going to allow more classical computation under dagger

    with dagger:
        a = 4
        theta = angle(1 / a)
        rx(q, theta)
        a /= 2
        phi = angle(1 / a)
        rz(q, phi)

invert_rotations.check()
```

> 

Reviewer: I don't think qasm should be necesary here. I think just the three backticks should be fine.

> ```qasm

Reviewer: Insert some text explanation here. It should be made clear that this is the order that Guppy does the gates and assignments.

a = 4;
theta = 1 / a;
a /= 2;
phi = 1 / a;
rz(-phi) q;
rx(-theta) q;
```

### Forbidden operations in dagger blocks

Dagger blocks have the same qubit-operation restrictions as control blocks and cannot contain control flow. They also cannot perform observable classical effects such as `output`, `panic`, or `exit`: reversing the quantum operations must not change when those effects occur.

```{code-cell} ipython3
---
tags: [raises-exception]
---
@guppy
def branch_in_dagger(q: qubit, flag: bool) -> None:
    with dagger:
        if flag:
            h(q)

branch_in_dagger.check()
```

```{code-cell} ipython3
---
tags: [raises-exception]
---
@guppy
def loop_in_dagger(q: qubit) -> None:
    with dagger:
        for _ in range(2):
            h(q)

loop_in_dagger.check()
```

```{code-cell} ipython3
---
tags: [raises-exception]
---
from guppylang.std.builtins import output

@guppy
def output_in_dagger(q: qubit) -> None:
    with dagger:
        output("value", True)

> output_in_dagger.check()

Reviewer: I think in rc2 this will give an error as expected.

```
### Combining dagger and controls

Modifiers compose. This example is a doubly controlled inverse of `h` followed by `s`:

```{code-cell} ipython3
@guppy
def doubly_controlled_inverse(c0: qubit, c1: qubit, q: qubit) -> None:
    with control(c0):
        with dagger:
            with control(c1):
                h(q)
                s(q)

doubly_controlled_inverse.check()
```

Resolving one modifier at a time gives this sequence:

```text
Source:              h q;                    s q;
Push control c1:     ctrl @ h c1, q;         ctrl @ s c1, q;
Resolve dagger:      ctrl @ sdg c1, q;      ctrl @ h c1, q;
Push control c0:     ctrl(2) @ sdg c0, c1, q;
                     ctrl(2) @ h c0, c1, q;
```

### Conjugation pattern

Many unitaries have a compute--action--uncompute form:

$$
U = V A V^\dagger.
$$

> The first part computes a basis change, `A` performs the action, and the final part uncomputes the basis change. This is a conjugation box.

Reviewer: ```suggestion
The first part computes a basis change, `A` performs the action, and the final part uncomputes the basis change. This is a conjugation pattern.
```

For a controlled conjugation, only the central action $A$ needs to be controlled:

$$
C[U] = (I \otimes V)\, C[A]\, (I \otimes V^\dagger).
$$

> This avoids controlling every operation in `V` and can substantially reduce the number of controlled, entangling gates.

Reviewer: ```suggestion
This avoids controlling every operation in $V$ and can substantially reduce the number of controlled, entangling gates.
```

```{code-cell} ipython3
from guppylang.std.quantum import rz

@guppy
def controlled_conjugation(c: qubit, q: qubit) -> None:
    # Compute V
    h(q)
    # Controlled action A
    with control(c):
        rz(q, angle(1 / 4))
    # Uncompute V†
    with dagger:
        h(q)

controlled_conjugation.check()
```

> #### Conjugation box with Pauli gadget

Reviewer: ```suggestion
#### Conjugation pattern with Pauli exponential
```

    with control(c), dagger:
        choose_gate(q, True)

modified_comptime_call.check()
```

Qubit allocation or measurement remains forbidden in a flagged compile-time function:

```{code-cell} ipython3
---
tags: [raises-exception]
---
@guppy.comptime(daggerable=True)
def allocating_comptime_function() -> None:
    q = qubit()
    h(q)
    measure(q)

allocating_comptime_function.compile()
```

### Conjugation pattern with functions

Now we can revisit the [previous example](#conjugation-box-with-pauli-gadget) using functions. 

```{code-cell} ipython3
@guppy(unitary=True)
def compute_zzyx_parity(qz0: qubit, qz1: qubit, qy: qubit, qx: qubit) -> None:
    rx(qy, angle(1 / 2))
    h(qx)
    cx(qz0, qx)
    cx(qz1, qx)
    cx(qy, qx)

@guppy(unitary=True)
def pauli_action(qx: qubit, theta: angle) -> None:
    rz(qx, theta)

@guppy
def controlled_pauli_zzyx_with_functions(
    c: qubit,
    qz0: qubit,
    qz1: qubit,
    qy: qubit,
    qx: qubit,
    theta: angle,
) -> None:
    compute_zzyx_parity(qz0, qz1, qy, qx)
    with control(c):
        pauli_action(qx, theta)
    with dagger:
        compute_zzyx_parity(qz0, qz1, qy, qx)

controlled_pauli_zzyx_with_functions.check()
```

> ## A complete example: Grover search

Reviewer: I think Grover's search makes more sense as a notebook example.
